ADO.NET/OAuthSampleApp/OAuthSampleApp.cs

```diff
@@ -0,0 +1,217 @@
+                }
+                catch (HttpRequestException hre)
+                {
+                    Console.WriteLine(hre.Message);
+                    Console.WriteLine("Refresh token invalid or expired. Attempting to get new access and refresh tokens");
+                    await DoPasswordGrant();
+                }
+
+                SetAccessToken();
+                retryCount++;
+            }
+        }
+    }
+
+    // password grant logs into the IDP using credentials in app.config
+    public static async Task DoPasswordGrant()
+    {
+        Dictionary<string, string> formData = new Dictionary<string, string>
+        {
+            {"grant_type",   "password"},
+            {"client_id",     ConfigurationManager.AppSettings["ClientId"]},
+            {"client_secret", ConfigurationManager.AppSettings["ClientSecret"]},
+            {"username",      ConfigurationManager.AppSettings["User"]},
+            {"password",      ConfigurationManager.AppSettings["Password"]}
+        };
+
+        await GetAndSetTokens(formData);
+    }
+
+    // refresh grant uses the refresh token to get a new access and refresh token
+    public static async Task DoTokenRefresh()
+    {
+        Dictionary<string, string> formData = new Dictionary<string, string>
+        {
+            {"grant_type",   "refresh_token"},
+            {"client_id",     ConfigurationManager.AppSettings["ClientId"]},
+            {"client_secret", ConfigurationManager.AppSettings["ClientSecret"]},
+            {"refresh_token", Environment.GetEnvironmentVariable(OAUTH_REFRESH_TOKEN_VAR_STRING, EnvironmentVariableTarget.User)}
+        };
+
+        await GetAndSetTokens(formData);
+    }
+
+    // uses the token url from the config to make an http request
+    // will always always set new access and refresh tokens, even if currently valid
+    private static async Task GetAndSetTokens(Dictionary<string, string> formData)
+    {
+        HttpClient httpClient = new HttpClient();
+
+        try
+        {
+            var content = new FormUrlEncodedContent(formData);
+
+            string tokenUrl = ConfigurationManager.AppSettings["TokenUrl"];
+            var response = await httpClient.PostAsync(tokenUrl, content);
+
+            response.EnsureSuccessStatusCode();
+
+            var result = await response.Content.ReadAsStringAsync();
+            JObject jObject = JObject.Parse(result);
+
+            // Set the access token as an environment variable
+            Environment.SetEnvironmentVariable(OAUTH_ACCESS_TOKEN_VAR_STRING, jObject["access_token"].ToString(), EnvironmentVariableTarget.User);
+
+            // Set the refresh token as an environment variable
+            Environment.SetEnvironmentVariable(OAUTH_REFRESH_TOKEN_VAR_STRING, jObject["refresh_token"].ToString(), EnvironmentVariableTarget.User);
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine("Error getting access token: " + ex.Message);

```

Comment:
This is true, but in the case of an expired refresh token the goal is also to acquire a new refresh token and not just an access token. So consider changing the error message. Maybe "error getting refresh and/or access tokens:" or something like that

Reply:
Good idea.